bip-vaults.mediawiki

```diff
@@ -0,0 +1,745 @@
+** If this value cannot be decoded as a CScriptNum and cast to an integer, script execution MUST fail and terminate immediately.
+** If this value is less than -1, script execution MUST fail and terminate immediately.
+** If this value is greater than or equal to 0, this spend is a recovery transaction and this value denotes the recovery output that corresponds to this vault input.
+* The parse of the other stack items depends on whether or not this is a recovery spend.
+
+==== <code>OP_VAULT</code> evaluation for recovery spend ====
+
+* If the recovery output does not have an <code>nValue</code> greater than this input's amount, the script MUST fail and terminate immediately.
+* (Deferred) if the recovery output does not have an <code>nValue</code> equal to the sum of all <code>OP_VAULT</code>/<code>OP_UNVAULT</code> inputs with a corresponding recovery sPK hash, the transaction validation MUST fail.<ref>'''How do recovery transactions pay for fees?''' If the recovery is unauthorized, fees are attached either via CPFP with an ephemeral anchor or as inputs which are solely spent to fees (i.e. no change output). If the recovery is authorized, fees can be attached in any manner, e.g. unrelated inputs and outputs or CPFP via anchor.</ref>
+** Note that in the draft implementation, this is facilitated by a "deferred check" which is queued by the script interpreter, but executed after the script interpreter has finished, in other validation code.<ref>'''Why does this proposal require a "deferred checks" framework for correct script evaluation?''' The deferred checks framework is an augmentation to execution of the Bitcoin script interpreter. Currently, the validity of each input is checked in an order-indepdendent manner across all inputs in a transaction. Because this proposal allows batching the spend of multiple vault inputs into a single recovery or withdrawal output, we need a mechanism to ensure that all expected values per output can be summed and then checked. This necessitates the introduction of an "aggregating" set of checks which can only be executed after each input's script is evaluated. Note that similar functionality would be required for batch input validation or cross-input signature aggregation.</ref>
+* The script must FAIL (by policy, not consensus) and terminate immediately if neither<ref>'''Why are recovery transactions required to be replaceable?''' In the case of unauthorized recoveries, an attacker may attempt to pin recovery transactions by broadcasting a "rebundled" version with a low fee rate. Vault owners must be able to overcome this with replacement. In the case of authorized recovery, if an attacker steals the recovery authorization key, the attacker may try to pin the recovery transaction during theft. Requiring replaceability ensures that the owner can always raise the fee rate of the recovery transaction, even if they are RBF rule #3 griefed in the process.</ref>
```

Comment:
The core (consensus-critical) specification should probably only include consensus rules and not recommended standardness/policy rules (which should perhaps not be included at all, just like they aren't in BIP341).

Reply:
I also didn't like the mixing of consensus and policy rules in the spec.  I think work on different layers has previously been dealt with by using different BIPs, e.g. BIPs 13 vs 16, or BIPs 141/143 vs 144.

That said, if separate BIPs is just too annoying, maybe the rules for policy could just be moved to a different section.

Reply:
I've moved this provision into a policy section (ca12c526ab11b0e7bf2b2da693405cf8ac350bfa), and will likely make recovery transaction structure (for unauthorized vs. authorized) a matter of policy as well, since the distinction only exists there to prevent pinning.

Reply:
Node policy is an individual decision and generally not a subject for BIPs at all.

Reply:
@luke-jr: Agreed but there are going to be future discussions on changing the default policy in Bitcoin Core and the recommended default policy in other implementations (which those implementations are free to ignore as are all users given it is only a policy default and not a consensus rule). It is inevitable with proposals for Lightning, vaults etc and the associated pinning vectors. Personally I would much rather any proposed changes to default policy were spec'ed and discussed in the open in the BIP process than just say merged into Core.

I do agree that mixing consensus and policy in the same proposal isn't ideal. Perhaps there can be a note to the reader on what the BIP author recommends for default policy changes associated with this proposal but making it clear that this BIP stands alone as a proposed consensus change and not a default policy change.